core/schema/modulesource.go

```diff
@@ -1,790 +1,998 @@
 	}
 
-	switch src.Kind {
+	switch parsedRef.kind {
 	case core.ModuleSourceKindLocal:
-		if filepath.IsAbs(parsed.modPath) {
-			cwdStat, err := bk.StatCallerHostPath(ctx, ".", true)
-			if err != nil {
-				return nil, fmt.Errorf("failed to stat caller's current working directory: %w", err)
-			}
-
-			relPath, err := pathutil.LexicalRelativePath(cwdStat.Path, parsed.modPath)
-			if err != nil {
-				return nil, err
-			}
-
-			parsed.modPath = relPath
+		inst, err = s.localModuleSource(ctx, query, bk, parsedRef.local.modPath, !args.DisableFindUp, args.AllowNotExists)
+		if err != nil {
+			return inst, err
 		}
-
-		src.AsLocalSource = dagql.NonNull(&core.LocalModuleSource{
-			RootSubpath: parsed.modPath,
-			RelHostPath: args.RelHostPath,
-		})
-
 	case core.ModuleSourceKindGit:
-		src.AsGitSource = dagql.NonNull(&core.GitModuleSource{})
-
-		src.AsGitSource.Value.Root = parsed.repoRoot.Root
-		src.AsGitSource.Value.HTMLRepoURL = parsed.repoRoot.Repo
-
-		// Determine usernames for source reference and actual cloning
-		sourceUser, cloneUser := parsed.sshusername, parsed.sshusername
-		if cloneUser == "" && parsed.scheme.IsSSH() {
-			cloneUser = "git"
+		inst, err = s.gitModuleSource(ctx, query, parsedRef.git, args.RefPin, !args.DisableFindUp)
+		if err != nil {
+			return inst, err
 		}
+	default:
+		return inst, fmt.Errorf("unknown module source kind: %s", parsedRef.kind)
+	}
 
-		if sourceUser != "" {
-			sourceUser += "@"
-		}
+	return inst, nil
+}
 
-		if cloneUser != "" {
-			cloneUser += "@"
-		}
+//nolint:gocyclo
+func (s *moduleSchema) localModuleSource(
+	ctx context.Context,
+	query dagql.Instance[*core.Query],
+	bk *buildkit.Client,
 
-		// Construct the source reference (preserves original input)
-		src.AsGitSource.Value.CloneRef = parsed.scheme.Prefix() + sourceUser + parsed.repoRoot.Root
+	// localPath is the path the user provided to load the module, it may be relative or absolute and
+	// may point to either the directory containing dagger.json or any subdirectory in the
+	// filetree under the directory containing dagger.json.
+	// When findUp is enabled, it can also be a name of a dependency in the default dagger.json found-up from the cwd.
+	localPath string,
 
-		// Construct the reference for actual cloning (ensures username for SSH)
-		cloneRef := parsed.scheme.Prefix() + cloneUser + parsed.repoRoot.Root
+	// whether to search up the directory tree for a dagger.json file. additionally, when enabled if a dagger.json is found-up
+	// and localPath is a named dependency in that dagger.json, the returned source will be for that dependency.
+	doFindUp bool,
 
-		subPath := "/"
-		if parsed.repoRootSubdir != "" {
-			subPath = parsed.repoRootSubdir
+	// if true, tolerate the localPath not existing on the filesystem (for dagger init on directories that don't exist yet)
+	allowNotExists bool,
+) (inst dagql.Instance[*core.ModuleSource], err error) {
+	if doFindUp {
```

Comment:
Should check if `localPath` exists first before doing `findUp` as a fallback:
```console
❯ dagger init --sdk=go graphql
Initialized module graphql in graphql
❯ dagger-dev shell --no-mod -c '.core | module-source graphql | as-string'
/Users/helder/Projects/dagger/modules/graphql
```
Expected: `/Users/helder/Projects/dagger/graphql`